src/highlight_processor.c

```diff
@@ -0,0 +1,200 @@
+#include "result_processor.h"
+#include "fragmenter.h"
+#include "value.h"
+#include "util/minmax.h"
+#include <ctype.h>
+
+/**
+ * Attempts to fragmentize a single field from its offset entries. This takes
+ * the field name, gets the matching field ID, retrieves the offset iterator
+ * for the field ID, and fragments the text based on the offsets. The fragmenter
+ * itself is in fragmenter.{c,h}
+ *
+ * Returns true if the fragmentation succeeded, false otherwise.
+ */
+static int fragmentizeOffsets(const RSSearchRequest *req, const char *fieldName,
+                              const char *fieldText, size_t fieldLen, RSIndexResult *indexResult,
+                              RSByteOffsets *byteOffsets, FragmentList *fragList) {
+  const FieldSpec *fs = IndexSpec_GetField(req->sctx->spec, fieldName, strlen(fieldName));
+  if (!fs) {
+    return 0;
+  }
+
+  int rc = 0;
+  RSOffsetIterator offsIter = RSIndexResult_IterateOffsets(indexResult);
+  FragmentTermIterator fragIter;
+  RSByteOffsetIterator bytesIter;
+  if (RSByteOffset_Iterate(byteOffsets, fs->id, &bytesIter) != REDISMODULE_OK) {
+    goto done;
+  }
+
+  FragmentTermIterator_InitOffsets(&fragIter, &bytesIter, &offsIter);
+  FragmentList_FragmentizeIter(fragList, fieldText, fieldLen, &fragIter);
+  rc = 1;
+
+done:
+  offsIter.Free(offsIter.ctx);
+  return rc;
+}
+
+#define HL_SEP_STR "... "
+
+// Strip spaces from a buffer in place. Returns the new length of the text,
+// with all duplicate spaces stripped and converted to a single ' '.
+static size_t stripDuplicateSpaces(char *s, size_t n) {
+  int isLastSpace = 0;
+  size_t oix = 0;
+  char *out = s;
+  for (size_t ii = 0; ii < n; ++ii) {
+    if (isspace(s[ii])) {
+      if (isLastSpace) {
+        continue;
+      } else {
+        isLastSpace = 1;
+        out[oix++] = ' ';
+      }
+    } else {
+      isLastSpace = 0;
+      out[oix++] = s[ii];
+    }
+  }
+  return oix;
+}
+
+static void summarizeField(const RSSearchRequest *req, const ReturnedField *fieldInfo,
```

Comment:
again, do not pass the request here. you only need the spec to pass to fragmentizeOffsets